### ISSUE TITLE
ruff vs. flake

### ISSUE THREAD

**Issue by rjschwei** [canonical/cloud-init#4339]
# Bug report
With e2ae75542f7 flake8 was removed in favor of ruff, but ruff is not as picky. For example newer versions of flake8 produce:

```
cloudinit/net/eni.py:84:16: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
cloudinit/net/eni.py:129:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
cloudinit/net/eni.py:147:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
cloudinit/net/network_state.py:562:39: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
cloudinit/net/network_state.py:585:20: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
cloudinit/sources/helpers/netlink.py:140:12: E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`
```

ruff accepts the existing syntax. Of course it is primarily cosmetic but that is why we run tools like flake, pylint and black. Also since we still support Python 3.6 we should use a version that is available for Python 3.6. On a system with Python 3.6:

```
-> tox -e ruff
GLOB sdist-make: /work/rjschwei/devel/cloud-init/setup.py
ruff create: /work/rjschwei/devel/cloud-init/.tox/ruff
ruff installdeps: ruff===0.0.282
ERROR: invocation failed (exit code 1), logfile: /work/rjschwei/devel/cloud-init/.tox/ruff/log/ruff-1.log
================================== log start ===================================
ERROR: Could not find a version that satisfies the requirement ruff===0.0.282 (from versions: 0.0.13, 0.0.14, 0.0.15, 0.0.16, 0.0.17)
ERROR: No matching distribution found for ruff===0.0.282

=================================== log end ====================================
ERROR: could not install deps [ruff===0.0.282]; v = InvocationError('/work/rjschwei/devel/cloud-init/.tox/ruff/bin/python -m pip install ruff===0.0.282', 1)
___________________________________ summary ____________________________________
ERROR:   ruff: could not install deps [ruff===0.0.282]; v = InvocationError('/work/rjschwei/devel/cloud-init/.tox/ruff/bin/python -m pip install ruff===0.0.282', 1)
```

## Steps to reproduce the problem
run newer versions of flake8 against current sources in main and compare with run of ruff. 

## Environment details
- Cloud-init version: 23.1
- Operating System Distribution: openSUSE Tumbleweed
- Cloud provider, platform or installer type:

**Comment by rjschwei** [comment]
We test on multiple python versions, currently `make flake8` is in our package builds [1], which is where this error showed up in openSUSE Tumbleweed where the default interpreter is Python 3.11. For openSUSE Leap and SLE the interpreter is Python 3.6. As such a `make ruff` target would fail with the version of ruff pinned to a version that is not available in Python 3.6. Meaning, if on my openSUSE Leap system I run `tox -e ruff` as shown in the original description I get the version error.

So my apologies for not expressing this correctly in the original description, albeit the problem still exists. Prior to the switch `tox -e flake8` worked, i.e. I could check my code before push, now I get the error in the description.

Ultimately I do not care for the tool, flake or ruff, I care for consistency. If we support Python 3.6 then one should be able to use Python 3.6 as an environment to submit code upstream and run the tests. If we do not do this then we  easily get into a "these tests work on Python 3.6 and these don't, but that OK" situation and that makes things really tricky, IMHO.


[1] https://build.opensuse.org/package/show/Cloud:Tools/cloud-init

**Comment by rjschwei** [comment]
@TheRealFalcon I think that is a good approach to cover the transition until we drop support for all Python versions where ruff doesn't work the way we want it to.
